bottlerocket-os/bottlerocket
# netdog: Enable conditional compilation based on network backend

**Issue number:**
Related to #2449 

**Description of changes:**
This set of temporary changes enables conditional compilation of `netdog` based on network backend.  "Temporary" in this case means that these changes (similar to those in #2741 ) will be removed once `systemd-networkd` development has finished.

A new function `emit_net_backend_cfgs()` has been added to `netdog`'s `build.rs`  that triggers off the `SYSTEMD_NETWORKD` environment variable that has been added to the build environment.  The crate can now use the following to conditionally compile code:
```
#[cfg(net_backend = "systemd-networkd")]
{ do stuff }
```

**Testing done:**
Added the following blurb to `netdog generate-net-config` to be conditionally compiled:
```
#[cfg(net_backend = "systemd-networkd")]
{
    eprintln!("I'm on systemd-networkd!!!")
}
```

* Build and boot an aws-k8s-1.22 image that runs properly and connects to the cluster - do not observe the above lines in the journal
* Build and locally boot an aws-dev variant. I added an early dependency on console-getty.service to the preconfigured.target since I knew early services would be broken but I still wanted to be able to poke around and verify the above. With the `systemd-networkd` build flag, I observed the above message in the journal.  Without the flag, the message didn't show up.
```
Starting Network Configuration...
[    2.081577] netdog[901]: I'm on systemd-networkd!!!
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
